EIPS/eip-5114.md

```diff
@@ -0,0 +1,66 @@
+---
+eip: 5114
+title: Soulbound Token
+description: A token that is attached to a "soul" at mint time and cannot be transferred after that.
+author: Micah Zoltu (@MicahZoltu)
+discussions-to: https://ethereum-magicians.org/t/eip-5114-soulbound-token/9417
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-05-30
+requires: 721
+---
+
+## Abstract
+A soulbound token is a token that is bound to another Non Fungible Token (NFT; e.g., a EIP-721 token) when it is minted, and cannot be transferred/moved after that.
+
+## Specification
+```solidity
+interface IERCXXXX {
+	// fired anytime a new instance of this token is minted
+	// this event **MUST NOT** be fired twice for the same `tokenId`
+	event Mint(uint256 indexed tokenId, address indexed nftAddress, uint256 indexed nftTokenId);
+
+	// returns the NFT token that owns this token.
+	// this function **MUST** throw if the token hasn't been minted yet
+	// this function **MUST** always return the same result every time it is called after it has been minted
+	// this function **MUST** return the same value as found in the original `Mint` event for the token
+	function ownerOf(uint256 index) external view returns (address nftAddress, uint256 nftTokenId);
+	
+	// returns a censorship resistant URI with details about this token collection
+	// the metadata returned by this is merged with the metadata return by `tokenUri(uint256)`
+	// the tokenUri **MUST** be immutable and content addressable (e.g., ipfs://)
+	// the tokenUri **MUST NOT** point at mutable/censorable content (e.g., https://)
```

Comment:
as much as I appreciate the sentiment, it seems out-of-scope to me to specify the nature of the storage, as it is outside the Ethereum network

the requirement for immutability is also questionable to me - I could see a use for a URI that can deliberately be changed or change itself over time

Reply:
If we don't constrain the storage, then the tokens functionally become mutable.  By asserting immutability throughout, we enable applications/users to build caching systems that they cannot built if we only encourage the behavior.

Reply:
Immutable data doesn't exist. From what I understand, only content-integrity assuring identifiers exist that can indicate mutation. Replication can be used to have non-mutated backups.

Reply:
I'm not following your claim here that immutable data doesn't exist.  If you ask IPFS for something by content hash, you will get either that exact thing, or you will get nothing (if it can't be found on the network).  There is no situation where you would get a different result from someone else or at a different point in time.

Regarding censorable content: I tend to agree that censorable content is fuzzy here, and I wish we could nail down something a bit more concrete.  What I want to avoid is people linking to content that a provider can remove at a future point in time.  The goal with these badges is that they are permanent, they aren't something that comes and goes over time.  If you link out to a traditional `https` resource, the host of that resource/owner of the domain can remove it.  Similarly, the domain may be seized and changed or taken away.  Something like IPFS on the other hand allows anyone to ensure the content survives, you don't have to rely on a single operator.

Reply:
"immutable data"/content-addressable criteria:

I agree with you that some networks of computers can create the illusion that some data is immutable. But on the basis of a principle, the data isn't immutable. E.g. on IPFS/Ethereum/Bitcoin when you run a full node, there's nothing stopping you from going into the chain data directory and switching a few bits. You can always mutate the data. It's likely though that the system maintaining the data will notice and e.g. notably fail or repair itself. That's what I meant by "Replication can be used to fallback on non-mutated backups (in a distributed system)."

How I'd describe it is that by requesting data using content-addressing, a user can guarantee to themselves that they'll either receive the exact data corresponding to the content-address, or nothing at all (e.g. when content-address and data don't align).

I think "content-addressing" is a better term than e.g. "immutable".

data availability guarantees/non-censorable:

I'd strongly opt for e.g. "highly-available data" rather than "non-censorable." From my perspective, censorship is about withholding information in the interest of a few. Data availability, on the other hand, is a term from the blockchain research world that describes the concept of persisting data within e.g. a p2p. E.g. all EVM storage has high data availability, and anything stored in the Bitcoin blockchain too.

For IPFS, actually, it doesn't have data availability built-in. To pin an IPFS content hash, one has to personally deploy a dedicated node that continues to make the content hash available within the distributed hash table. IMO, if you wanted the ultimate data availability for Soulbound Badges, it'd probably be best to store the contents of `tokenUri` on-chain as e.g. a struct. I'm not kidding. It's because if you outsource delivering extremely high-quality data availability to other networks not aligned with Ethereum (e.g. IPFS), it may be that one day they just go down/stop working while Ethereum still works.

Now talking about storing the badge's metadata on-chain, I think it's obligatory I bring up https://github.com/ethereum/EIPs/issues/735

Reply:
Lets move this conversation over to the discussions-to link as I'm hoping this will get merged as a draft soon and I don't want this conversation to get lost.